File: Try_GameEngine5/src/com/example/try_gameengine/action/MovementActionDecoratorFactory.java
package com.example.try_gameengine.action;

import java.util.ArrayList;
import java.util.List;

public class MovementActionDecoratorFactory {
	
	public static List<Class<? extends MovementDecorator>> createDHMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(HalfDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createDHDMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(HalfDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createDDMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(DoubleDecorator.class);
//		decoratorClassList.add(HalfDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createDDDDMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createCopyMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(CopyMoveDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createCCMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(CopyMoveDecorator.class);
		decoratorClassList.add(CopyMoveDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createCDMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(CopyMoveDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createDCMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(CopyMoveDecorator.class);
		return decoratorClassList;
	}
	
	public static List<Class<? extends MovementDecorator>> createCDCMovementDecorator(){
		List<Class<? extends MovementDecorator>> decoratorClassList = new ArrayList<Class<? extends MovementDecorator>>();
		decoratorClassList.add(CopyMoveDecorator.class);
		decoratorClassList.add(DoubleDecorator.class);
		decoratorClassList.add(CopyMoveDecorator.class);
		return decoratorClassList;
	}
	
}

File: Try_GameEngine5/src/com/example/try_gameengine/test/MovementActionDecoratorFactory.kt
package com.example.try_gameengine.test

import com.example.try_gameengine.action.CopyMoveDecorator
import com.example.try_gameengine.action.DoubleDecorator
import com.example.try_gameengine.action.HalfDecorator
import com.example.try_gameengine.action.MovementDecorator

object MovementActionDecoratorFactory {
    fun createDHMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(HalfDecorator::class.java)
        return decoratorClassList
    }

    fun createDHDMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(HalfDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        return decoratorClassList
    }

    fun createDDMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(DoubleDecorator::class.java)
        //		decoratorClassList.add(HalfDecorator.class);
        decoratorClassList.add(DoubleDecorator::class.java)
        return decoratorClassList
    }

    fun createDDDDMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        return decoratorClassList
    }

    fun createCopyMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(CopyMoveDecorator::class.java)
        return decoratorClassList
    }

    fun createCCMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(CopyMoveDecorator::class.java)
        decoratorClassList.add(CopyMoveDecorator::class.java)
        return decoratorClassList
    }

    fun createCDMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(CopyMoveDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        return decoratorClassList
    }

    fun createDCMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(CopyMoveDecorator::class.java)
        return decoratorClassList
    }

    fun createCDCMovementDecorator(): MutableList<Class<out MovementDecorator?>?> {
        val decoratorClassList: MutableList<Class<out MovementDecorator?>?> =
            ArrayList<Class<out MovementDecorator?>?>()
        decoratorClassList.add(CopyMoveDecorator::class.java)
        decoratorClassList.add(DoubleDecorator::class.java)
        decoratorClassList.add(CopyMoveDecorator::class.java)
        return decoratorClassList
    }
}
